addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java

```diff
@@ -566,110 +656,126 @@ private void readliness(String terminator) throws IOException {
-                        } catch (Exception e) {
-                            logger.error("An exception occurred while calling the DeviceStatusListener", e);
-                            unregisterDeviceStatusListener(deviceStatusListener);
-                        }
+
+                c = DeviceConfiguration.create(di);
+                configurations.add(c);
+                c.setRoomId(di.getRoomId());
+                String roomName = "";
+                for (RoomInformation room : msg.rooms) {
+                    if (room.getPosition() == di.getRoomId()) {
+                        roomName = room.getName();
                     }
                 }
-            } else if (message.getType() == MessageType.L) {
-                ((L_Message) message).updateDevices(devices, configurations);
-                logger.trace("{} devices found.", devices.size());
-            } else if (message.getType() == MessageType.S) {
-                dutyCycle = ((S_Message) message).getDutyCycle();
-                freeMemorySlots = ((S_Message) message).getFreeMemorySlots();
-                updateCubeState();
-                if (((S_Message) message).isCommandDiscarded()) {
-                    logger.warn("Last Send Command discarded. Duty Cycle: {}, Free Memory Slots: {}", dutyCycle,
-                            freeMemorySlots);
-                } else {
-                    logger.debug("S message. Duty Cycle: {}, Free Memory Slots: {}", dutyCycle, freeMemorySlots);
+                c.setRoomName(roomName);
+            }
+        } else if (message.getType() == MessageType.C) {
+            DeviceConfiguration c = null;
+            for (DeviceConfiguration conf : configurations) {
+                if (conf.getSerialNumber().equalsIgnoreCase(((C_Message) message).getSerialNumber())) {
+                    c = conf;
+                    break;
+                }
+            }
+
+            if (c == null) {
+                configurations.add(DeviceConfiguration.create(message));
+            } else {
+                c.setValues((C_Message) message);
+                Device di = getDevice(((C_Message) message).getSerialNumber());
+                if (di != null) {
+                    di.setProperties(((C_Message) message).getProperties());
+                    ;
                 }
-            } else if (message.getType() == MessageType.N) {
-                if (((N_Message) message).getRfAddress() != null) {
-                    newInclusionDeviceFound((N_Message) message);
+            }
+            if (exclusive == true) {
+                for (DeviceStatusListener deviceStatusListener : deviceStatusListeners) {
+                    try {
+                        Device di = getDevice(((C_Message) message).getSerialNumber());
+                        if (di != null) {
+                            deviceStatusListener.onDeviceConfigUpdate(getThing(), di);
+                        }
+                    } catch (NullPointerException e) {
```

Comment:
wasn't my code. I was afraid to refactor it, because I didn't know what exactly could NPE...